File: Android project/AndroidNetworkniginJava/app/src/main/java/com/mirkamol/androidnetworkniginjava/network/volley/VolleyHttp.java
package com.mirkamol.androidnetworkniginjava.network.volley;

import androidx.annotation.Nullable;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.toolbox.StringRequest;
import com.mirkamol.androidnetworkniginjava.model.Poster;

import org.json.JSONObject;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class VolleyHttp {

    public static boolean IS_TESTER = true;
    public static String SERVER_DEVELOPMENT = "https://jsonplaceholder.typicode.com/";
    public static String SERVER_PRODUCTION = "https://jsonplaceholder.typicode.com/";

    public static String server(String url) {
        if (IS_TESTER)
            return SERVER_DEVELOPMENT + url;
        return SERVER_PRODUCTION + url;
    }

    public static HashMap<String, String> headers() {
        HashMap<String, String> headers = new HashMap<>();
        headers.put("Content-type", "application/json; charset=UTF-8");
        return headers;
    }

    public static void get(String api, HashMap<String, String> params, VolleyHandler volleyHandler) {
        StringRequest stringRequest = new StringRequest(Request.Method.GET, server(api),
                (Response.Listener<String>) response -> {
                    volleyHandler.onSuccess(response);
                }, error -> {
            volleyHandler.onError(error.toString());
        }) {
            @Nullable
            @Override
            protected Map<String, String> getParams() throws AuthFailureError {
                return params;
            }
        };

        MyApplication.instance.addToRequestQueue(stringRequest);
    }

    public static void post(String api, HashMap<String, Object> body, VolleyHandler volleyHandler) {
        StringRequest stringRequest = new StringRequest(Request.Method.POST, server(api),
                (Response.Listener<String>) response -> {
                    volleyHandler.onSuccess(response);
                }, error -> {
            volleyHandler.onError(error.toString());
        }) {
            @Override
            public byte[] getBody() throws AuthFailureError {
                return new JSONObject(body).toString().getBytes(StandardCharsets.UTF_8);
            }

            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                return headers();
            }
        };

        MyApplication.instance.addToRequestQueue(stringRequest);
    }

    public static void put(String api, HashMap<String, Object> body, VolleyHandler volleyHandler) {
        StringRequest stringRequest = new StringRequest(Request.Method.PUT, server(api),
                (Response.Listener<String>) response -> {
                    volleyHandler.onSuccess(response);
                }, error -> {
            volleyHandler.onError(error.toString());
        }) {
            @Override
            public byte[] getBody() throws AuthFailureError {
                return new JSONObject(body).toString().getBytes(StandardCharsets.UTF_8);
            }

            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                return headers();
            }
        };

        MyApplication.instance.addToRequestQueue(stringRequest);
    }

    public static void deletePost(String url, VolleyHandler volleyHandler) {
        StringRequest stringRequest = new StringRequest(Request.Method.DELETE, server(url),
                (Response.Listener<String>) response -> {
                    volleyHandler.onSuccess(response);
                }, error -> {
            volleyHandler.onError(error.toString());
        });
        MyApplication.instance.addToRequestQueue(stringRequest);

    }

    public static String API_LIST_POST = "posts";
    public static String API_SINGLE_POST = "posts/"; //id
    public static String API_CREATE_POST = "posts";
    public static String API_UPDATE_POST = "posts/"; //id
    public static String API_DELETE_POST = "posts/"; //id

    public static HashMap<String, String> paramsEmpty() {
        return new HashMap<>();
    }

    public static HashMap<String, Object> paramsCreate(Poster post) {
        HashMap<String, Object> params = new HashMap<>();
        params.put("title", post.title);
        params.put("body", post.body);
        params.put("userId", post.userId);
        return params;
    }

    public static HashMap<String, Object> paramsUpdate(Poster post) {
        HashMap<String, Object> params = new HashMap<>();
        params.put("id", post.id);
        params.put("title", post.title);
        params.put("body", post.body);
        params.put("userId", post.userId);
        return params;
    }
}

File: Android project/Lesson6VolleyNetWorking/app/src/main/java/com/mirkamol/lesson6volleynetworking/network/VolleyHttp.kt
package com.mirkamol.lesson6volleynetworking.network

import com.android.volley.Response
import com.android.volley.toolbox.StringRequest
import com.mirkamol.lesson6volleynetworking.activities.MyApplication
import com.mirkamol.lesson6volleynetworking.model.Data
import com.mirkamol.lesson6volleynetworking.network.volley.VolleyHandler
import com.mirkamol.lesson6volleynetworking.utils.Logger
import org.json.JSONObject


class VolleyHttp {

    companion object{
        val TAG = MyApplication::class.java.simpleName
        val IS_TESTER = true
        val SERVER_DEVELOPMENT ="http://dummy.restapiexample.com/api/v1/"
        val SERVER_PRODUCTION = "http://dummy.restapiexample.com/api/v1/"

        fun server(url:String):String{
            return if (IS_TESTER){
                SERVER_DEVELOPMENT + url
            }else{
                SERVER_PRODUCTION + url
            }
        }

        fun headers(): HashMap<String, String> {
            val headers = HashMap<String, String>()
            headers["Content-type"] = "application/json; charset=UTF-8"
            return headers
        }

        fun get(api: String, params: HashMap<String, String>, volleyHandler: VolleyHandler) {
            val stringRequest = object : StringRequest(
                Method.GET, server(api),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.e(TAG, error.toString())
                    volleyHandler.onError(error.toString())
                }) {
                override fun getParams(): MutableMap<String, String> {
                    return params
                }
            }
            MyApplication.instance!!.addToRequestQueue(stringRequest)
        }
        // get single post
        fun getSinglePost(api: String, params: HashMap<String, String>, volleyHandler: VolleyHandler) {
            val stringRequest = object : StringRequest(
                Method.GET, server(api),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.e(TAG, error.toString())
                    volleyHandler.onError(error.toString())
                }) {
                override fun getParams(): MutableMap<String, String> {
                    return params
                }
            }
            MyApplication.instance!!.addToRequestQueue(stringRequest)
        }

        fun post(api: String, body: HashMap<String, Any>, volleyHandler: VolleyHandler) {
            val stringRequest = object : StringRequest(
                Method.POST, server(api),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.d(TAG, error.toString())
                    volleyHandler.onError(error.toString())
                }) {
                override fun getHeaders(): MutableMap<String, String> {
                    return headers()
                }

                override fun getBody(): ByteArray {
                    return JSONObject(body as Map<*, *>).toString().toByteArray()
                }
            }
            MyApplication.instance!!.addToRequestQueue(stringRequest)
        }

        fun put(api: String, body: HashMap<String, Any>, volleyHandler: VolleyHandler) {
            val stringRequest = object : StringRequest(
                Method.PUT, server(api),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.d(TAG, error.toString())
                    volleyHandler.onError(error.toString())
                }) {
                override fun getHeaders(): MutableMap<String, String> {
                    return headers()
                }

                override fun getBody(): ByteArray {
                    return JSONObject(body as Map<*, *>).toString().toByteArray()
                }
            }
            MyApplication.instance!!.addToRequestQueue(stringRequest)
        }

        fun delete(url: String, volleyHandler: VolleyHandler) {
            val stringRequest = object : StringRequest(
                Method.DELETE, server(url),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.d(TAG, error.toString())
                    volleyHandler.onError(error.toString())
                }) {
            }
            MyApplication.instance!!.addToRequestQueue(stringRequest)
        }

        /**
         *  Request Api`s
         */
        var API_LIST_POST = "employees"
        var API_SINGLE_POST = "employee/" //id
        var API_CREATE_POST = "create"
        var API_UPDATE_POST = "update/" //id
        var API_DELETE_POST = "delete/" //id

        /**
         *  Request Param`s
         */



        fun paramsEmpty(): HashMap<String, String> {
            return HashMap<String, String>()
        }

        fun paramsCreate(data: Data): HashMap<String, Any> {
            val params = HashMap<String, Any>()
            params["employee_name"] = data.employee_name
            params["employee_salary"] = data.employee_salary
            params["employee_age"] = data.employee_age
            params["profile_image"] = data.profile_image
            return params
        }

        fun paramsUpdate(data: Data): HashMap<String, Any> {
            val params = HashMap<String, Any>()
            params["employee_name"] = data.employee_name
            params["employee_salary"] = data.employee_salary
            params["employee_age"] = data.employee_age
            params["profile_image"] = data.profile_image
            return params
        }

        /**
         *  Response Parse`s
         */
    }
}